source/common/stats/stats_matcher_impl.cc

```diff
@@ -38,19 +45,74 @@ StatsMatcherImpl::StatsMatcherImpl(const envoy::config::metrics::v3::StatsConfig
   }
 }
 
-bool StatsMatcherImpl::rejects(const std::string& name) const {
-  //
-  //  is_inclusive_ | match | return
-  // ---------------+-------+--------
-  //        T       |   T   |   T     Default-inclusive and matching an (exclusion) matcher, deny.
-  //        T       |   F   |   F     Otherwise, allow.
-  //        F       |   T   |   F     Default-exclusive and matching an (inclusion) matcher, allow.
-  //        F       |   F   |   T     Otherwise, deny.
-  //
-  // This is an XNOR, which can be evaluated by checking for equality.
-
-  return (is_inclusive_ == std::any_of(matchers_.begin(), matchers_.end(),
-                                       [&name](auto& matcher) { return matcher.match(name); }));
+// If the last string-matcher added is a case-sensitive prefix match, and the
+// prefix ends in ".", then this drops that match and adds it to a list of
+// prefixes. This is beneficial because token prefixes can be handled more
+// efficiently as a StatName without requiring conversion to a string.
+//
+// In the future, other matcher patterns could be optimized in a similar way,
+// such as:
+//   * suffixes that begin with "."
+//   * exact-matches
+//   * substrings that begin and end with "."
+//
+// These are left unoptimized for the moment to keep the code-change simpler,
+// and because we haven't observed an acute performance need to optimize those
+// other patterns yet.
+void StatsMatcherImpl::optimizeLastMatcher() {
+  std::string prefix;
+  if (matchers_.back().getCaseSensitivePrefixMatch(prefix) && absl::EndsWith(prefix, ".") &&
+      prefix.size() > 1) {
+    prefixes_.push_back(stat_name_pool_->add(prefix.substr(0, prefix.size() - 1)));
+    matchers_.pop_back();
+  }
+}
+
+StatsMatcher::FastResult StatsMatcherImpl::fastRejects(StatName stat_name) const {
+  if (rejectsAll()) {
+    return FastResult::NoMatch;
```

Comment:
Can’t this return “rejects”?

Reply:
yes, good catch. that was working before I switched to enums, but there was not a test that discerned the difference between fast-reject and slow-reject of this case. I added a couple of tests for this.